tests/unit/backend/services/mail.service.spec.ts

```diff
@@ -981,4 +983,169 @@ describe('mail.service', () => {
+          },
+          MOCK_BOUNCE_TYPE,
+        ),
+        headers: {
+          // Hardcode in tests in case something changes this.
+          'X-Formsg-Email-Type': 'Admin (bounce notification)',
+          'X-Formsg-Form-ID': MOCK_FORM_ID,
+          'X-Formsg-Submission-ID': MOCK_SUBMISSION_ID,
+        },
+      }
+    }
+
+    let expectedArgs: MailOptions
+    beforeAll(async () => {
+      expectedArgs = await generateExpectedArg()
+    })
+
+    it('should send bounce notification successfully', async () => {
+      // Arrange
+      // sendMail should return mocked success response
+      const mockedResponse = 'mockedSuccessResponse'
+      sendMailSpy.mockResolvedValueOnce(mockedResponse)
+
+      // Act
+      const sent = await mailService.sendBounceNotification({
+        emailRecipients: MOCK_RECIPIENTS,
+        bounceType: MOCK_BOUNCE_TYPE,
+        bouncedRecipients: MOCK_BOUNCED_EMAILS,
+        formId: MOCK_FORM_ID,
+        formTitle: MOCK_FORM_TITLE,
+        submissionId: MOCK_SUBMISSION_ID,
+      })
+
+      // Assert
+      expect(sent).toEqual(mockedResponse)
+      // Check arguments passed to sendNodeMail
+      expect(sendMailSpy).toHaveBeenCalledTimes(1)
+      expect(sendMailSpy).toHaveBeenCalledWith(expectedArgs)
+    })
+
+    it('should reject with error when email is invalid', async () => {
+      // Arrange
+      const invalidEmail = 'notAnEmail'
+
+      // Act
+      const sent = mailService.sendBounceNotification({
+        emailRecipients: [invalidEmail],
+        bounceType: MOCK_BOUNCE_TYPE,
+        bouncedRecipients: MOCK_BOUNCED_EMAILS,
+        formId: MOCK_FORM_ID,
+        formTitle: MOCK_FORM_TITLE,
+        submissionId: MOCK_SUBMISSION_ID,
+      })
+
+      // Assert
+      await expect(sent).rejects.toThrowError('Invalid email error')
+    })
+
+    it('should autoretry when 4xx error is thrown by sendNodeMail and pass if second try passes', async () => {
+      // Arrange
+      // sendMail should return mocked success response
+      const mockedResponse = 'mockedSuccessResponse'
+      const mock4xxReject = {
+        responseCode: 454,
+        message: 'oh no something went wrong',
+      }
+      sendMailSpy
+        .mockRejectedValueOnce(mock4xxReject)
+        .mockReturnValueOnce(mockedResponse)
+
+      // Act
+      const sent = await mailService.sendBounceNotification({
+        emailRecipients: MOCK_RECIPIENTS,
+        bounceType: MOCK_BOUNCE_TYPE,
+        bouncedRecipients: MOCK_BOUNCED_EMAILS,
+        formId: MOCK_FORM_ID,
+        formTitle: MOCK_FORM_TITLE,
+        submissionId: MOCK_SUBMISSION_ID,
+      })
+
+      // Assert
+      expect(sent).toEqual(mockedResponse)
+      // Check arguments passed to sendNodeMail
+      // Should have been called two times since it rejected the first one and
+      // resolved
+      expect(sendMailSpy).toHaveBeenCalledTimes(2)
+      expect(sendMailSpy).toHaveBeenCalledWith(expectedArgs)
```

Comment:
For completeness, how about checking the expectedArgs for both calls:

```
expect(sendMailSpy).toHaveBeenNthCalledWith(1, expectedArgs)
expect(sendMailSpy).toHaveBeenNthCalledWith(2, expectedArgs)
```

Reply:
awesome, will do!